### ISSUE TITLE
[FEATURE] Support for version 1.4 schema

### ISSUE THREAD

**Issue by madpah** [CycloneDX/cyclonedx-python-lib#67]
Creating issue to track adoption of the forthcoming version 1.4 schema specification for CycloneDX.

Preview work can be seen [here](https://github.com/CycloneDX/specification/blob/v1.4-dev/schema), but this IS NOT FINAL.

Current estimates are that version 1.4 will be finalised late 2021, early 2022.

**Comment by jkowalleck** [comment]
lets postpone this feature, until 1.4 is through the RFC phase. 
maybe extend the on-hold until 1.4 got finalized

**Comment by madpah** [comment]
1.4 is now in draft with only minor fixes now expected before it is finalised and released early 2022.

See https://github.com/CycloneDX/specification/tree/v1.4-dev

**Comment by madpah** [comment]
Work has commenced in branch **feat/67-schema-version-1.4-xml**

**Comment by madpah** [comment]
Work items:

- [ ] Update all schemas in `schemas` folder
- [ ] Add XML schema validation in unit tests only (this prevents `lxml` from becoming a mainstream dependency)
- [ ] Add JSON schema validation in unit tests only (prevent dependency bloat)
- [ ] Schema Change: Add support for `bom.metadata.tools.tool.externalReferences`
- [ ] Schema Change: Make `bom.component.version` optional
- [ ] Schema Change: Add support for `bom.component.releaseNotes`
- [ ] Schema Change: Vulnerabilities are now part of the core schema in 1.4 - `bom.vulnerabilities`
- [ ] Schema Change: Support for `$schema` in JSON format #104
- [ ] Schema Change: JSON schema in 1.4 has been hardened - JSF added

Future work, not specific to version 1.4:
- [ ] Add support for `bom.services`
- [ ] Add support for `bom.externalReferences`
- [ ] Add support for `bom.dependencies`
- [ ] Add support for `bom.compositions`
- [ ] Add support for `bom.properties`

@jkowalleck - appreciate your quick review on the above (first) list of key items to get done for 1.4 spec.

**Comment by jkowalleck** [comment]
@madpah love the list you accumulated. 

some important mentions/additions/summaries/annotations: 
* "Vulnerabilities are now part of the core schema in 1.4"
  * _cyclonedx-python-lib_ has an implementation for it already. regarding spec < 1.4 
  * so the serializers need to implement the CDX-builtin vuln for JSON & XML for spec > 1.4  
    and may fallback to the existing implementation for XML spec < 1.4  
    and omit vulns for JSON spec < 1.4 since
    @madpah i think this would a noticable challende that would require more work than the other transitions.
* "JSON schema in 1.4 has been hardened " - this means that json exensions are no longer possible with spec 1.4.   
  see https://github.com/CycloneDX/specification/blob/v1.4-dev/schema/bom-1.4-SNAPSHOT.schema.json*   
  `"additionalProperties": false,` everywhere - nothing but the content of the spec is allowed,
  - this should be no issue, since the schema extention never existed for JSON (https://github.com/CycloneDX/specification/blob/v1.4-dev/schema/ext/vulnerability-1.0.xsd)
* component's `version` is now optional - no longer in `required`
  - so the data models need to reflect this change. 
  - it was agreed to set `version` to an empty string in spec<1.4 for backwards compatibility.

---
@madpah feel free to edit this message, to add more implementation details we need to keep track of

**Comment by madpah** [comment]
Thanks @jkowalleck - agree with your summary - and thanks for pointing out the agreement to set `version == ''` - had not spotted that!

**Comment by jkowalleck** [comment]
the thing that worries me the most: 
schema 1.4 is not relaxed - is "hardened" - see https://github.com/CycloneDX/cyclonedx-python-lib/issues/67#issuecomment-1000182082 
this means there must be no properties in the JSON that are not in the spec - so we should add a validator to the library,
so down-streams/users of the lib can assert that the resulting JSON is valid.

**Comment by madpah** [comment]
Understand @jkowalleck - I've already added XML and JSON schema validation for unit tests. But we can easily add that to the main code.

I was avoiding XML schema validation in the main code as that would require `lxml`, which is a bit bloaty and platform specific. Thoughts?

**Comment by jkowalleck** [comment]
good point, @madpah .

lets have the validators for internal unit-tests for now and get experienced with it.
there must be multiple xml- and json schema validator libraries out there - so we need to find one that suites us the most.

we might need to file a decision paper ([like this one](https://github.com/CycloneDX/cyclonedx-php-library/blob/master/docs/dev/decisions/JsonSchemaValidator.md))
and gather non-functional requirements/constraints, 
before we can make a validator public.

if we like the solution we are using for our own tests, we might make it public later.

**Comment by jkowalleck** [comment]
@madpah when downloading/shipping the spec1.4 json schema, we need to have the JSF schema also.
JSF it is used in spec 1.4 json schema as a relative reference - just like the SPDX schema.

we might not have proper support for JSF implemented, but for schema validation (in unit tests) we need the file.

**Comment by madpah** [comment]
Thanks for the heads up @jkowalleck - haven't got to JSON for 1.4 yet :-)

**Comment by madpah** [comment]
Pre-release published to PyPi: https://pypi.org/project/cyclonedx-python-lib/1.0.0rc0/

FYI @jkowalleck

**Comment by jkowalleck** [comment]
will do a review of https://github.com/CycloneDX/cyclonedx-python-lib/pull/108

**Comment by madpah** [comment]
Released!

Thanks @jkowalleck.
